CMake: Add comments why we need URL downloads of external projects

uklotzde opened the pull request.

uklotzde committed:
CMake: Use Git for downloading external projects

uklotzde commented:
Not sure if this breaks the Debian/Ubuntu builds: https://github.com/mixxxdj/mixxx/pull/3776#discussion_r611213871

It shouldn't. Manually downloading the sources and placing them in the download directory should always work. We already do this for Fedora.

daschuer commented:
I get now obscure error messages, but no  significant status message for a hint how to treat this situation: 
```
[ 15%] Built target gmock_main
Hinweis: Checke '0.15.1' aus.

Sie befinden sich im Zustand eines 'losgelösten HEAD'. Sie können sich
umschauen, experimentelle Änderungen vornehmen und diese committen, und
Sie können alle möglichen Commits, die Sie in diesem Zustand machen,
ohne Auswirkungen auf irgendeinen Branch verwerfen, indem Sie einen
weiteren Checkout durchführen.

Wenn Sie einen neuen Branch erstellen möchten, um Ihre erstellten Commits
zu behalten, können Sie das (jetzt oder später) durch einen weiteren Checkout
mit der Option -b tun. Beispiel:

  git checkout -b <neuer-Branchname>

HEAD ist jetzt bei 90068a2 Merge pull request #55 from xsco/fix/cmake-sqlite-amalgamation-scope
cd /home/daniel/workspace/qt5mixxx/build/Debug/libdjinterop-prefix/src && /usr/bin/cmake -E touch /home/daniel/workspace/qt5mixxx/build/Debug/libdjinterop-prefix/src/libdjinterop-stamp/libdjinterop-download
/usr/bin/cmake -E echo_append
```

daschuer commented on CMakeLists.txt:
```diff
@@ -1724,8 +1724,8 @@ if(ENGINEPRIME)
     set(DJINTEROP_LIBRARY "lib/${CMAKE_STATIC_LIBRARY_PREFIX}djinterop${CMAKE_STATIC_LIBRARY_SUFFIX}")
     file(MAKE_DIRECTORY "${CMAKE_CURRENT_BINARY_DIR}/download")
```
Is this directory still used?

uklotzde replied:
> I get now obscure error messages, but no significant status message for a hint how to treat this situation:
> 

The detached checkout warnings appear always. Don't know how to suppress them.

uklotzde replied:
The DOWNLOAD_DIR is used in the same ways as before.

daschuer commented:
In case of no internet connection I get: 
```
make[2]: Verzeichnis „/home/daniel/workspace/qt5mixxx/build/Debug“ wird verlassen
fatal: unable to access 'https://github.com/xsco/libdjinterop/': Could not resolve host: github.com
-- Had to git clone more than once:
          3 times.
CMake Error at /home/daniel/workspace/qt5mixxx/build/Debug/libdjinterop-prefix/tmp/libdjinterop-gitclone.cmake:31 (message):
  Failed to clone repository: 'https://github.com/xsco/libdjinterop'
CMakeFiles/libdjinterop.dir/build.make:87: recipe for target 'libdjinterop-prefix/src/libdjinterop-stamp/libdjinterop-download' failed
make[2]: *** [libdjinterop-prefix/src/libdjinterop-stamp/libdjinterop-download] Error 1
```

This appears in the middle of the build and not at the beginning where the other requirements are checked.

daschuer commented:
I can confirm It works on Ubuntu Bionic using https://github.com/daschuer/mixxx/tree/2.4.bionic, but the log output can be removed. I guess it is now quite hard to fix an issue, if you don't know insides.

mr-smidge commented:
> The detached checkout warnings appear always. Don't know how to suppress them.

Does enabling `GIT_SHALLOW TRUE` suppress the message?

daschuer commented:
The warning itself alone is not a problem. It is the combination with no other status messages that the user can understand why it is there. 

https://cmake.org/cmake/help/latest/module/ExternalProject.html
> If GIT_SHALLOW is enabled then GIT_TAG works only with branch names and tags. A commit hash is not allowed.

I guess fro our release we will use a tagged version, right?

mr-smidge commented:
> I guess fro our release we will use a tagged version, right?

I can't speak for keyfinder, but as maintainer of libdjinterop my policy is to only ever have applications depend on a proper, tagged version.  I also ensure that any distro packages (e.g. Ubuntu PPA) align with the same tag.

Using `GIT_SHALLOW` still permits temporarily pointing at a feature branch of keyfinder/libdjinterop by name during development, which is useful.

uklotzde committed:
Use GIT_SHALLOW when downloading external projects

daschuer commented on CMakeLists.txt:
```diff
@@ -1831,8 +1832,9 @@ if(KEYFINDER)
+      GIT_REPOSITORY "https://github.com/mixxxdj/libkeyfinder"
+      GIT_TAG "v2.2.4"
+      GIT_SHALLOW TRUE
       DOWNLOAD_DIR "${CMAKE_CURRENT_BINARY_DIR}/download/libkeyfinder"
```
This line can be removed. I have just tested that it works without.

daschuer reviewed (changes requested):
The SHALLOW flag does the trick. Please remove the download folder and we can merge.

uklotzde commented:
I first need to check if this doesn't break the offline builds!

uklotzde commented:
Offline builds doesn't seem to work with GIT_REPOSITORY/GIT_TAG and without DOWNLOAD_DIR. We should document that and keep the URL download.

uklotzde committed:
Download external project sources by URL instead of Git

Revert the previous commits that turned out as inappropriate.

https://github.com/mixxxdj/mixxx/pull/3779

uklotzde commented:
Reverted and comments added.

daschuer reviewed (approved):
LGTM. Do you want to keep the intermediate commits in history? 
I think we should squash this PR, what do you think?
